ext/socket/lib/socket.rb

```diff
@@ -624,8 +648,487 @@ def accept_nonblock(exception: true)
+              connecting_sockets.delete connectable_socket
+              connectable_sockets.each do |other_connectable_socket|
+                other_connectable_socket.close unless other_connectable_socket.closed?
+              end
+              state = :success
+              break
+            else
+              failed_ai = connecting_sockets.delete connectable_socket
+              inspected_ip_address = failed_ai.ipv6? ? "[#{failed_ai.ip_address}]" : failed_ai.ip_address
+              last_error = SystemCallError.new("connect(2) for #{inspected_ip_address}:#{failed_ai.ip_port}", sockopt.int)
+              connectable_socket.close unless connectable_socket.closed?
+
+              next if connectable_sockets.any?
+
+              if selectable_addrinfos.empty? && connecting_sockets.empty? && hostname_resolution_queue.empty?
+                if !hostname_resolution_queue.closed? && !wait_for_hostname_resolution_patiently
+                  wait_for_hostname_resolution_patiently = true
+                  connection_attempt_delay_expires_at = current_clocktime + PATIENTLY_RESOLUTION_DELAY
+                else
+                  state = :failure
+                end
+              elsif selectable_addrinfos.any?
+                # Wait for connection attempt delay timeout in next loop
+              else
+                # Wait for connection to be established or hostname resolution in next loop
+                connection_attempt_delay_expires_at = nil
+              end
+            end
+          end
+        elsif hostname_resolved&.any?
+          family_name, res = hostname_resolution_queue.get
+          selectable_addrinfos.add(family_name, res) unless res.is_a? Exception
+          connection_attempt_delay_expires_at = nil if wait_for_hostname_resolution_patiently
+          state = :v46w
+        else
+          if selectable_addrinfos.empty? && connecting_sockets.empty? && hostname_resolution_queue.empty?
+            state = :failure
+          elsif selectable_addrinfos.any?
+            state = :v46c
+          else
+            connection_attempt_delay_expires_at = nil
+          end
+        end
+
+        next
+      when :success
+        break connected_socket
+      when :failure
+        raise last_error
+      when :timeout
+        raise Errno::ETIMEDOUT, 'user specified timeout'
+      end
+    end
+
+    if block_given?
+      begin
+        yield ret
+      ensure
+        ret.close
+      end
+    else
+      ret
+    end
+  ensure
+    if fast_fallback
+      hostname_resolution_threads.each do |thread|
+        thread&.exit
+      end
+
+      hostname_resolution_queue&.close_all
+
+      connecting_sockets.each do |connecting_socket|
+        connecting_socket.close unless connecting_socket.closed?
+      end
+    end
+  end
+
+  def self.specified_family_name_and_next_state(hostname)
+    if    hostname.match?(/:/)                             then [:ipv6, :v6c]
```

Comment:
This is just a corner case, but several platforms (at least Linux and macOS) accept a colon (:) in hostnames, and DNS also allows a domain name with a colon. For instance, `Socket.tcp("x::y", 80)` and `Socket.tcp("1::1::1", 80)` specify hostnames (not IPv6 address literals), in which case both of the IPv4 and IPv6 name resolutions should be tried; otherwise, they fail to connect if the names only have IPv4 addresses.
So I think here we need the exact pattern to match IPv6 address literals (including link-local addresses with scope). False negatives (redundant name resolutions) are better than false positives (connection failures).